checker/cost.go

```diff
@@ -0,0 +1,430 @@
+package checker
+
+import (
+	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"math"
+
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/overloads"
+)
+
+// CostEstimator estimates the sizes of variable length input data and the costs of functions.
+type CostEstimator interface {
+	// EstimateSize returns a SizeEstimate for the given AstNode, or nil if
+	// the estimator has no estimate to provide. The size is equivalent to the result of the CEL `size()` function:
+	// length of strings and bytes, number of map entries or number of list items.
+	// EstimateSize is only called for AstNodes where
+	// CEL does not know the size; EstimateSize is not called for values defined inline in CEL where the size
+	// is already obvious to CEL.
+	EstimateSize(element AstNode) *SizeEstimate
+	// EstimateCallCost returns the estimated cost of an invocation, or nil if
+	// the estimator has no estimate to provide.
+	EstimateCallCost(overloadId string, target *AstNode, args []AstNode) *CostEstimate
+}
+
+// AstNode represents an AST node for the purpose of cost estimations.
+type AstNode interface {
+	// Path returns a path to the AstNode. The first path element is a variable. All subsequent path elements are
+	// field selectors or '*' to indicate traversal into the elements or  of a map of list.
+	Path() []string
```

Comment:
It's possible you'll want to consider using the `AttributePattern` in the future as it will give you features like `input.*.field` which may be desirable for handling matches against element access while still retaining fine-grained estimates for sub-element access. 

Using `AttributePattern` would require this code be moved into another package though, probably something like `cel/internal` (new).

Reply:
Sounds good. I've added some test around edge cases where I need to distinguish between the map key type and value type, so I landed on using "@items", "@keys" and "@values" instead of "*" in the path.